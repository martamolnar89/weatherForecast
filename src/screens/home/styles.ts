import { StyleSheet, Dimensions } from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    minHeight: Dimensions.get('screen').height - 60,
  },
  loading: {
    justifyContent: 'center',
    paddingHorizontal: 20,
    paddingVertical: 20,
  },
  header: {
    paddingVertical: 10,
    paddingHorizontal: 20,
    backgroundColor: 'lightgray',
  },
  dataContainer: {
    paddingHorizontal: 20,
  },
  data: {
    flexDirection: 'row',
    marginHorizontal: 20,
  },
  tempData: {
    flexDirection: 'column',
    marginHorizontal: 20,
    marginVertical: 5,
  },
  title: {
    fontSize: 20,
    marginVertical: 10,
  },
  error: {
    justifyContent: 'flex-start',
    paddingVertical: 20,
    paddingHorizontal: 20,
  },
});

export default styles;
