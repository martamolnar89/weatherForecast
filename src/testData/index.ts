const testData = {
  cod: '200',
  message: 0,
  cnt: 40,
  list: [
    {
      dt: 1574953200,
      main: {
        temp: 3.57,
        temp_min: 3.57,
        temp_max: 6.16,
        pressure: 1003,
        sea_level: 1003,
        grnd_level: 995,
        humidity: 81,
        temp_kf: -2.59,
      },
      weather: [
        {
          id: 500,
          main: 'Rain',
          description: 'light rain',
          icon: '10n',
        },
      ],
      clouds: {
        all: 100,
      },
      wind: {
        speed: 3.71,
        deg: 28,
      },
      sys: {
        pod: 'n',
      },
      dt_txt: '2019-11-28 15:00:00',
    },
    {
      dt: 1574964000,
      main: {
        temp: 5.14,
        temp_min: 5.14,
        temp_max: 7.08,
        pressure: 1003,
        sea_level: 1003,
        grnd_level: 996,
        humidity: 76,
        temp_kf: -1.94,
      },
      weather: [
        {
          id: 803,
          main: 'Clouds',
          description: 'broken clouds',
          icon: '04d',
        },
      ],
      clouds: {
        all: 80,
      },
      wind: {
        speed: 3.03,
        deg: 37,
      },
      sys: {
        pod: 'd',
      },
      dt_txt: '2019-11-28 18:00:00',
    },
  ],
  city: {
    id: 5391959,
    name: 'San Francisco',
    coord: {
      lat: 37.7749,
      lon: -122.4195,
    },
    country: 'US',
    population: 805235,
    timezone: -28800,
    sunrise: 1574953410,
    sunset: 1574988729,
  },
};

export default testData;
