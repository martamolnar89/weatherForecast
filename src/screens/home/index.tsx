import React, {
  useState,
  useEffect,
  FunctionComponent,
  useCallback,
} from 'react';
import {
  View,
  Text,
  Image,
  ActivityIndicator,
  ScrollView,
  RefreshControl,
} from 'react-native';
import { connect } from 'react-redux';
import Geolocation from 'react-native-geolocation-service';
import styles from './styles';
import { Forecast, RootState } from 'types/index';
import { getForecastData } from 'redux/actions/forecast';
import { wait } from 'utils/helpers';

interface HomeScreenProps {
  weatherData: Forecast | null;
  fetchErrorMessage: string | null;
}

function getForecastByCurrentPosition() {
  return new Promise((resolve, reject) =>
    Geolocation.getCurrentPosition(
      async position => {
        await getForecastData(
          position.coords.latitude,
          position.coords.longitude,
        );
        resolve();
      },
      error => {
        reject(error);
      },
      { enableHighAccuracy: true, timeout: 15000, maximumAge: 10000 },
    ),
  );
}

const HomeScreen: FunctionComponent<HomeScreenProps> = ({
  weatherData,
  fetchErrorMessage,
}: HomeScreenProps) => {
  const [loading, setLoading] = useState(true);
  const [locationErrorMessage, setLocationErrorMessage] = useState<
    string | null
  >(null);
  const [refreshing, setRefreshing] = useState(false);

  useEffect(() => {
    const refreshData = async () => {
      try {
        await getForecastByCurrentPosition();
      } catch (error) {
        setLocationErrorMessage(
          'Permission to access location was denied, weather forecast is unavailable!',
        );
      }
      setLoading(false);
    };
    refreshData();
  }, []);

  const onRefresh = useCallback(async () => {
    setRefreshing(true);
    try {
      await getForecastByCurrentPosition();
    } catch (error) {
      console.warn(error);
    }
    await wait(2000);
    setRefreshing(false);
  }, [refreshing]);

  return (
    <ScrollView
      style={styles.container}
      contentInsetAdjustmentBehavior="automatic"
      refreshControl={
        <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
      }>
      {loading && (
        <View style={styles.loading}>
          <ActivityIndicator size="large" color="gray" />
        </View>
      )}
      {locationErrorMessage && (
        <View style={styles.error}>
          <Text style={styles.title}>{locationErrorMessage}</Text>
        </View>
      )}
      {fetchErrorMessage && (
        <View testID="fetchErrorView" style={styles.error}>
          <Text style={styles.title}>{fetchErrorMessage}</Text>
        </View>
      )}
      {weatherData && weatherData.city && (
        <View testID="header" style={styles.header}>
          <Text style={styles.title}>
            {weatherData.city.name} weather forecast
          </Text>
        </View>
      )}
      {weatherData &&
        weatherData.list &&
        weatherData.list.map(data => (
          <View key={data.dt} style={styles.dataContainer}>
            <Text style={styles.title}>{data.dt_txt}</Text>
            <View style={styles.data}>
              <Image
                style={{ width: 50, height: 50 }}
                source={{
                  uri: `http://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`,
                }}
              />
              <Text style={styles.title}>{data.main.temp} &#x2103; </Text>
              <Text style={styles.title}>{data.weather[0].main}</Text>
              <View style={styles.tempData}>
                <Text>min: {data.main.temp_min} &#x2103; </Text>
                <Text>max: {data.main.temp_max} &#x2103;</Text>
              </View>
            </View>
          </View>
        ))}
    </ScrollView>
  );
};

const mapStateToProps = (state: RootState) => ({
  weatherData: state.forecast.weatherData,
  fetchErrorMessage: state.forecast.errorMessage,
});

export const HomeScreenComponent = HomeScreen;
export default connect(mapStateToProps, {})(HomeScreen);
