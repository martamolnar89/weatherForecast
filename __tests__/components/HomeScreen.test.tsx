import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { HomeScreenComponent as HomeScreen } from '../../src/screens/home/index';
import testData from '../../src/testData/index';

describe('HomeScreen', () => {
  test('initial render', () => {
    const homeScreen = renderer.create(
      <HomeScreen weatherData={testData} fetchErrorMessage={null} />,
    );

    expect(homeScreen).toMatchSnapshot('HomeScreen');
  });

  test('initial render with fetch error message', () => {
    const homeScreen = renderer.create(
      <HomeScreen weatherData={null} fetchErrorMessage={'FAILED'} />,
    );

    expect(homeScreen).toMatchSnapshot('HomeScreenFetchError');
  });

  test('render header', () => {
    const homeScreen = renderer.create(
      <HomeScreen weatherData={testData} fetchErrorMessage={'FAILED'} />,
    );

    const homeScreenInstance = homeScreen.root;

    expect(homeScreenInstance.findByProps({ testID: 'header' })).not.toBe(null);
  });

  test('render fetch error message when fetch weather forecast failed', () => {
    const homeScreen = renderer.create(
      <HomeScreen weatherData={null} fetchErrorMessage={'FAILED'} />,
    );

    const homeScreenInstance = homeScreen.root;

    expect(
      homeScreenInstance.findByProps({ testID: 'fetchErrorView' }),
    ).not.toBe(null);
  });
});
