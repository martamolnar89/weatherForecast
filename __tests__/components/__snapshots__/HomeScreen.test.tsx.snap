// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen initial render with fetch error message: HomeScreenFetchError 1`] = `
<RCTScrollView
  contentInsetAdjustmentBehavior="automatic"
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  style={
    Object {
      "flex": 1,
      "minHeight": 1274,
    }
  }
>
  <RCTRefreshControl />
  <View>
    <View
      style={
        Object {
          "justifyContent": "center",
          "paddingHorizontal": 20,
          "paddingVertical": 20,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="gray"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
    <View
      style={
        Object {
          "justifyContent": "flex-start",
          "paddingHorizontal": 20,
          "paddingVertical": 20,
        }
      }
      testID="fetchErrorView"
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "marginVertical": 10,
          }
        }
      >
        FAILED
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`HomeScreen initial render: HomeScreen 1`] = `
<RCTScrollView
  contentInsetAdjustmentBehavior="automatic"
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  style={
    Object {
      "flex": 1,
      "minHeight": 1274,
    }
  }
>
  <RCTRefreshControl />
  <View>
    <View
      style={
        Object {
          "justifyContent": "center",
          "paddingHorizontal": 20,
          "paddingVertical": 20,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="gray"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "lightgray",
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="header"
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "marginVertical": 10,
          }
        }
      >
        San Francisco
         weather forecast
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "marginVertical": 10,
          }
        }
      >
        2019-11-28 15:00:00
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginHorizontal": 20,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "http://openweathermap.org/img/wn/10n@2x.png",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginVertical": 10,
            }
          }
        >
          3.57
           ℃ 
        </Text>
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginVertical": 10,
            }
          }
        >
          Rain
        </Text>
        <View
          style={
            Object {
              "flexDirection": "column",
              "marginHorizontal": 20,
              "marginVertical": 5,
            }
          }
        >
          <Text>
            min: 
            3.57
             ℃ 
          </Text>
          <Text>
            max: 
            6.16
             ℃
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "marginVertical": 10,
          }
        }
      >
        2019-11-28 18:00:00
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginHorizontal": 20,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "http://openweathermap.org/img/wn/04d@2x.png",
            }
          }
          style={
            Object {
              "height": 50,
              "width": 50,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginVertical": 10,
            }
          }
        >
          5.14
           ℃ 
        </Text>
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginVertical": 10,
            }
          }
        >
          Clouds
        </Text>
        <View
          style={
            Object {
              "flexDirection": "column",
              "marginHorizontal": 20,
              "marginVertical": 5,
            }
          }
        >
          <Text>
            min: 
            5.14
             ℃ 
          </Text>
          <Text>
            max: 
            7.08
             ℃
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
